packages/repository/src/repositories/relation.factory.ts

```diff
@@ -0,0 +1,164 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/example-todo
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {EntityCrudRepository} from '.';
+import {
+  Model,
+  RelationType,
+  AnyObject,
+  Options,
+  Entity,
+  Class,
+  FilterBuilder,
+  Filter,
+  WhereBuilder,
+  Where,
+  DataObject,
+} from '..';
+import {cloneDeep, isArray} from 'lodash';
+import {
+  HasManyEntityCrudRepository,
+  DefaultHasManyEntityCrudRepository,
+} from './relation.repository';
+
+export type RelationDefinitionBase = {
+  type: RelationType;
+  modelFrom: Class<Entity> | string;
+  keyTo: string;
+  keyFrom: string;
+};
+
+export interface HasManyDefinition extends RelationDefinitionBase {
+  type: RelationType.hasMany;
+}
+/**
+ * Enforces a constraint on a repository based on a relationship contract
+ * between models. Returns a relational repository that exposes applicable CRUD
+ * method APIs for the related target repository. For example, if a Customer model is
+ * related to an Order model via a HasMany relation, then, the relational
+ * repository returned by this method would be constrained by a Customer model
+ * instance's id(s).
+ *
+ * @param constraint The constraint to apply to the target repository. For
+ * example, {id: '5'}.
+ * @param relationMetadata The relation metadata used to used to describe the
+ * relationship and determine how to apply the constraint.
+ * @param targetRepository The repository which represents the target model of a
+ * relation attached to a datasource.
+ *
+ */
+export function constrainedRepositoryFactory<T extends Entity, ID>(
+  constraint: AnyObject,
```

Comment:
IIUC: when creating a HasMany repository for a given customer, we are passing a constrain `{id:1}` which contains a subset of customer (modelFrom) properties. This is not clear to me, why not to pass order (modelTo) properties instead?

Since our HasMany relation is always using a primary key (possibly a composite key), I think we should rename this argument name to better convey our intent - pass the single primary key or all parts of the composite key that's used to identify the source model.

Reply:
Sounds good to me. I wanted to keep it generic, but as the first step of implementing `HasMany` relation, it should be specific to it. I think this aligns with @raymondfeng's view as well https://github.com/strongloop/loopback-next/pull/1342#discussion_r189048398.

Reply:
HasMany is NOT always using the primary key. Any unique keys should be fine. Maybe something like `referencedKeyValue` make sense as we typically define a FK as:
```sql
ALTER TABLE Orders
ADD FOREIGN KEY (CustomerId) REFERENCES Customer(id);
```

Reply:
`id properties` form the primary key for our models' `*byId()` operations. They can be just unique keys in the database. They are logic PKs.